enforcer/datapath_proxy.go

```diff
@@ -0,0 +1,543 @@
+	ip = addr.data[2:6]
+	port := uint16(int(addr.data[0])<<8 + int(addr.data[1]))
+
+	return ip, port, nil
+}
+
+// Initiate the downstream connection
+func (p *Proxy) downConnection(ip []byte, port uint16) (int, error) {
+
+	fmt.Println("Dialing connection to backend:", net.IPv4(ip[0], ip[1], ip[2], ip[3]).To4(), port)
+
+	var err error
+	fd, err := syscall.Socket(syscall.AF_INET, syscall.SOCK_STREAM, 0)
+	if err != nil {
+		zap.L().Error("Socket create failed", zap.String("Error", err.Error()))
+	}
+
+	err = syscall.SetsockoptInt(fd, syscall.SOL_SOCKET, syscall.SO_MARK, proxyMarkInt)
+	if err != nil {
+		zap.L().Error("Sockopt  failed", zap.String("Error", err.Error()))
+	}
+	address := &syscall.SockaddrInet4{
+		Port: int(port),
+	}
+	copy(address.Addr[:], ip)
+
+	if p.Encrypt && p.Forward {
+		// config, err := p.loadTLS()
+		// if err != nil {
+		// 	return nil, err
+		// }
+
+		// downConn, err = tls.Dial("tcp", backend, config)
+		// if err != nil {
+		// 	return nil, err
+		// }
+	} else {
+		err = syscall.Connect(fd, address)
+		if err != nil {
+			zap.L().Error("Connect Error", zap.String("Connect Error", err.Error()))
+			return fd, err
+		}
+
+		addr, _ := syscall.Getpeername(fd)
+		remote := addr.(*syscall.SockaddrInet4)
+		addr, _ = syscall.Getsockname(fd)
+		local := addr.(*syscall.SockaddrInet4)
+
+		conntrackHdl := conntrack.NewHandle()
+		if connterror := conntrackHdl.ConntrackTableUpdateMark(net.IPv4(local.Addr[0], local.Addr[1], local.Addr[2], local.Addr[3]).String(),
+			net.IPv4(remote.Addr[0], remote.Addr[1], remote.Addr[2], remote.Addr[3]).String(),
+			syscall.IPPROTO_TCP,
+			uint16(local.Port),
+			uint16(remote.Port),
+			constants.DefaultConnMark,
+		); connterror != nil {
+			zap.L().Error("Unable to mark flow")
+		}
+
+	}
+
+	return fd, nil
+}
+
+//CompleteEndPointAuthorization -- Aporeto Handshake on top of a completed connection
+//We will define states here equivalent to SYN_SENT AND SYN_RECEIVED
+func (p *Proxy) CompleteEndPointAuthorization(backendip string, backendport uint16, upConn net.Conn, downConn int, contextID string) error {
+	puContext, err := p.datapath.contextTracker.Get(contextID)
+	if err != nil {
+		zap.L().Error("Did not find context")
+	}
+	puContext.(*PUContext).Lock()
+	defer puContext.(*PUContext).Unlock()
+	pu := puContext.(*PUContext)
+	//addr := upConn.RemoteAddr().String()
+
+	if pu.PUType == constants.LinuxProcessPU {
+		//Are we client or server proxy
+
+		if len(puContext.(*PUContext).Ports) > 0 && puContext.(*PUContext).Ports[0] != "0" {
+
+			return p.StartServerAuthStateMachine(backendip, backendport, upConn, downConn, contextID)
+		}
+		//We are client no advertised port
+		return p.StartClientAuthStateMachine(backendip, backendport, upConn, downConn, contextID)
+
+	}
+	//Assumption within a container two applications talking to each other won't be proxied.
+	//If backend ip is non local we are client else we are server
+	islocalIP := func() bool {
+		for _, ip := range p.IPList {
```

Comment:
I am not sure this is needed. But even if its, should we use a map of IPs and not a loop ?